PEP 604 - Updates and Edits

MaggieMoss opened the pull request:
- Added additional examples and details to the specification section.


cc: @gvanrossum @pprados

hugovk commented on pep-0604.rst:
```diff
@@ -8,6 +8,7 @@ Type: Standards Track
 Content-Type: text/x-rst
 Created: 28-Aug-2019
 Python-Version: 3.10
+Discussions-To: typing-sig@python.org
```
The CI failed because the headers are out of order: `Discussions-To` should follow `BDFL-Delegate`:

https://github.com/python/peps/blob/82a3d1f3dfd3ee293d722131e4bb7d4de4d551e2/pep-0001.txt#L512-L532

MaggieMoss replied:
Thank you!

gvanrossum commented on pep-0604.rst:
```diff
@@ -73,13 +74,59 @@ Here are some examples of what we can do with this feature.
+    # typing.List[typing.Union[str, int]]
+    typing.List[str | int]
+
+    # list[typing.Union[str, int]] (Currently TypeError: 'type' object has no attribute '__getitem__')
```
I don't think that error is relevant. PEP 585 makes it work starting in 3.9. You might merge this with the previous example like you did for dict/Dict below.

gvanrossum commented on pep-0604.rst:
```diff
@@ -73,13 +74,59 @@ Here are some examples of what we can do with this feature.
+    typing.Dict[str, int | float]
+    dict[str, int | float]
+
+The new syntax should be accepted for calls to `isinstance` as long as the Union items are valid arguments to `isinstance` themselves.
```
Use double backticks (also in various places below).

```suggestion
The new syntax should be accepted for calls to ``isinstance`` as long as the Union items are valid arguments to ``isinstance`` themselves.
```

gvanrossum commented on pep-0604.rst:
```diff
@@ -73,13 +74,59 @@ Here are some examples of what we can do with this feature.
+
+The new syntax should be accepted for calls to `isinstance` as long as the Union items are valid arguments to `isinstance` themselves.
+
+isinstance::
```
You need a blank line hereafter for the markup to work. (Also below.)

gvanrossum commented on pep-0604.rst:
```diff
@@ -73,13 +74,59 @@ Here are some examples of what we can do with this feature.
+
+  # invalid
+
+The existing `typing.Union` and `|` syntax should be equal.::
```
equal -> equivalent? Also drop the period before the colon (again below).

gvanrossum commented on pep-0604.rst:
```diff
@@ -73,13 +74,59 @@ Here are some examples of what we can do with this feature.
+  None | t == typing.Optional[t]
+
+A new Union.__str__() method should be implemented.::
+  (int | list[str]).__str__()
```
I would spell this using either `str(...)` or `print(...)`. But actually, it should be a new `__repr__` -- by defining just `__repr__` both `str()` and `repr()` will produce the same output. (And the REPL uses `repr()`.)

MaggieMoss commented:
@gvanrossum Thank you so much for the feedback, I really appreciate it. 

I've implemented your suggestions. Happy to continue iterating :) 


> Also, let's get approval from @pprados to add your name to the PEP authors.

Apologies, I thought you were cc'd on the email where he mentioned he was okay with this. I removed my name and I'll let @pprados confirm here.

gvanrossum commented on pep-0604.rst:
```diff
@@ -74,61 +76,72 @@ Simplified Syntax
     # Instead of typing.Dict[str, typing.Union[int, float]]
-    typing.Dict[str, int | float]
     dict[str, int | float]
+    list[str | int]
```
Looks like this line belongs in the previous group, and you accidentally dropped `typing.Dict[str, int | float]` here.

gvanrossum commented on pep-0604.rst:
```diff
@@ -74,61 +76,72 @@ Simplified Syntax
     assert str | int == Union[str,int]
     assert str | int | float == Union[str, int, float]
```
I think the call example and the asserts are unnecessary (what the asserts say is repeated better below).

gvanrossum commented on pep-0604.rst:
```diff
@@ -74,61 +76,72 @@ Simplified Syntax
+
+  # invalid
+  assert issubclass(bool, bool | list[int])
+
```
There's a lot of duplication for isinstance and subclass -- can you combine the two sections?

gvanrossum commented on pep-0604.rst:
```diff
@@ -74,61 +76,72 @@ Simplified Syntax
+  assert issubclass(bool, bool | list[int])
+
 
 Once the Python language is extended, mypy [1]_ and other type checkers will
 need to be updated to accept this new syntax.
```
This seems to belong in the Implementation section below. You could also combine that section with the Reference Implementation section further below.

MaggieMoss commented:
Thanks again for the review. :)

gvanrossum commented on pep-0604.rst:
```diff
@@ -130,30 +128,12 @@ The new syntax should be accepted for calls to ``isinstance`` as long as the Uni
-Once the Python language is extended, mypy [1]_ and other type checkers will
-need to be updated to accept this new syntax.
+    # valid
+    assert issubclass(bool, int | float)
```
```suggestion
    issubclass(bool, int | float)
```

gvanrossum commented on pep-0604.rst:
```diff
@@ -130,30 +128,12 @@ The new syntax should be accepted for calls to ``isinstance`` as long as the Uni
-Implementation
-==============
+    # invalid
+    assert issubclass(bool, bool | list[int])
```
```suggestion
    issubclass(bool, bool | list[int])
```

gvanrossum commented on pep-0604.rst:
```diff
@@ -123,7 +169,7 @@ For more details about discussions, see links below:
 
-- This syntax can be more readable, and is similary to others languages (Scala, ...)
+- This syntax can be more readable, and is similar to other languages (Scala, ...)
 - At runtime, ``int|str`` might return a simple object in 3.9, rather than everything that
```
```suggestion
- At runtime, ``int|str`` might return a simple object in 3.10, rather than everything that
```

gvanrossum commented on pep-0604.rst:
```diff
@@ -123,7 +169,7 @@ For more details about discussions, see links below:
 
```
The headline just above here says `Union[type1|type2]`. I think that `|` is a mistake and the intention was to write `Union[type1, type2]`.

gvanrossum commented on pep-0604.rst:
```diff
@@ -163,16 +209,26 @@ if we accept to not be compatible with the dynamic evaluation of annotations (``
 
 - If they were permitted, then instance checking could use an extremely clean-looking notation
-- The implementation can use the tuple present in ``Union`` parameter, without create a new instance
+- The implementation can use the tuple present in ``Union`` parameter, without creating a new instance
```
In #1540, @methane proposes to strike this sentence, as it is not a real example. (Probably `int|float|str` is actually slower than `(int, float, str)`, which maybe should be listed under CONS.)

gvanrossum commented on pep-0604.rst:
```diff
@@ -163,16 +209,26 @@ if we accept to not be compatible with the dynamic evaluation of annotations (``
 CONS:
 
-- Must migrate all the ``typing`` module in ``builtin``
+- Must migrate all of the ``typing`` module in ``builtin``
```
This is not correct (as your implementation proves :-).

gvanrossum commented on pep-0604.rst:
```diff
@@ -118,13 +164,13 @@ For more details about discussions, see links below:
 
-1. Add a new operator for ``Union[type1|type2]``?
+1. Add a new operator for ``Union[type1, type2]``?
 -------------------------------------------------
```
```suggestion
--------------------------------------------------
```
